cirq/ops/projectors.py

```diff
@@ -0,0 +1,75 @@
+from typing import Any, cast, Dict, List, Tuple, TYPE_CHECKING, Union
+
+import numpy as np
+
+from cirq import value
+from cirq.ops import raw_types
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality
+class Projector(raw_types.Gate):
+    """A non-unitary gate that projects onto a single qubit."""
+
+    def __init__(self,
+                 projection_basis: Union[List[List[float]], np.ndarray],
+                 qid_shape: Tuple[int, ...] = (2,)):
+        """
+        Args:
+            projection_basis: a (2**num_qubits, p) matrix that lists the
+                projection vectors.
+            qid_shape: Specifies the dimension of the projection. The default is
+                2 (qubit).
+
+        Raises:
+            ValueError: If the basis vector is empty.
+        """
+        if type(projection_basis) == np.ndarray:
+            projection_array = cast(np.ndarray, projection_basis)
+        else:
+            projection_array = np.array(projection_basis)
+
+        if np.prod(qid_shape) != projection_array.shape[1]:
+            raise ValueError(
+                "Invalid shape " +
+                f"{np.array(projection_array).shape} for qid_shape {qid_shape}")
+        self._projection_basis = projection_array
+        self._qid_shape = qid_shape
+
+    def _projection_basis_(self):
+        return self._projection_basis.tolist()
+
+    def _qid_shape_(self) -> Tuple[int, ...]:
+        return self._qid_shape
+
+    def _has_unitary_(self) -> bool:
+        return False
+
+    def _is_parameterized_(self) -> bool:
+        return False
+
+    def _channel_(self):
+        A = self._projection_basis
+        AH = np.transpose(np.conjugate(A))
+        result = AH * np.linalg.inv(np.matmul(A, AH)) * A
+        return (result,)
```

Comment:
Does this break if I give an overcomplete basis ?
```python
A = [[1,0], [0,1], [1,1]]
```
Should we care about such a case @viathor ?

Reply:
Yes it does break. I guess We could do an SVD instead, but maybe it's better to let the user fix it?

Leaving unresolved since @viathor hasn't chimed in.

Regardless, I added a test for it.